### ISSUE TITLE
Setting to deactivate "All Exceptions" breaking for first few seconds of a debug session

### ISSUE THREAD

**Issue by justingrant** [microsoft/vscode-js-debug#837]
Feature Request: could we add a `delayAllExceptionsBreakpointsMsecs: number` setting to launch.json for node debugging?  If the "All Exceptions" breakpoint setting is enabled and this setting is present, VSCode would *not* break on exceptions within the first N msecs of a debug session.  Note that this setting shouldn't affect uncaught exceptions, only caught exceptions if the "All Exceptions" breakpoint setting is enabled. 

**Justification**

Many JavaScript frameworks catch and handle all thrown exceptions so that they can display custom error UI. This makes the "uncaught exceptions" breakpoint useless. You need to use "All Exceptions" breakpoints instead.  But then you run into 10-100 "normal exceptions" when starting every debug session, because many libraries intentionally throw exceptions as part of their startup process. For example: 
* The json3 library, used all over the place (e.g. webpack dev server), always throws 2+ exceptions at startup while checking browser compatibility.
* Facebook's JS SDK also throws exceptions on startup when checking for browser JSON parsing support. 
* react-select, a common dropdown-UI library, always throws an exception at startup, also to check browser compatibility
* react itself throws exceptions for warnings about various components not following best practices.
* Node's dynamic resolver code throws exceptions when a module isn't loaded yet, and catches the exception to load the module. 
* The bluebird promise library (used by npm CLI) throws a few exceptions as part of its startup

I could keep going, but you get the idea. What all these have in common is that they happen in the first few seconds of a debug session, before any of the code runs that I'm actually trying to debug!

If we had the ability to simply not break during those first few seconds, it'd save me 1+ minutes at the start of every debug session!

BTW, I know about the the "skipFiles" option 
(https://code.visualstudio.com/docs/nodejs/nodejs-debugging#_skipping-uninteresting-code-node-chrome) which is somewhat helpful, but this feature has a few problems: 
* Exceptions continue to propagate up the stack. The file may be ignored, but not the exception. So, for example, instead of the IDE breaking at the call site in the json3 library, it will break somewhere in a huge transpiled webpack bundle file, which takes 10+ seconds to load into the editor because the file is so huge.
* File-level granularity is too coarse. Ignoring entire files might be risky because then I could miss real exceptions later. For example, there's no way that I'd want to ignore "real" exceptions from react or from the Facebook SDK!
* For scripts loaded over HTTP (not from disk), like the FB SDK or a webpack bundle file, you can't specify the entire URL path, only the filename part (e.g. "sdk.js" or "bundle.js"). Woe to you if you have two downloaded SDKs both called "sdk.js" !

A more granular alternative solution, like the "anti-breakpoint" proposal at Microsoft/vscode#53795, would also solve this problem, but I assume it'd be much harder to implement than an exception delay-- both on the VSCode side and from users. Also, a problem with this approach is that it wouldn't work for bundle files where the row/col would change from one debug session to the next.

**Comment by connor4312** [comment]
I understand the sentiment but I think a delay is not the best approach. What I would usually want is some way to tell the debugger to start catching exceptions at a certain point in the program execution. https://github.com/microsoft/vscode/issues/102181 proposes a greater degree of runtime participation from the debugger, along those lines you could imagine there can be some function call (not sure how it would be fired or what the name is yet) like `vscodeEnableExceptions() -> Promise` that would turn on handling uncaught exceptions at this moment in the execution.

**Comment by justingrant** [comment]
> you could imagine there can be some function call (not sure how it would be fired or what the name is yet) like `vscodeEnableExceptions() -> Promise` that would turn on handling uncaught exceptions at this moment in the execution.

Hmm. Often when you're debugging, you're not in control of much of the underlying code that you're debugging. Requiring a code change may be problematic, esp. if you're debugging sourcemapped code where the original source can't be easily edited during a debug session.

I'm not opposed to a "start debugging now!" API, but I also think that a "no code change" solution is needed too.

**Comment by connor4312** [comment]
You could also in principle have that function call in a conditional breakpoint (e.g. `vscodeEnableExceptions(); return false`) to avoid a code change, though that is somewhat more obscure.

Generally when I would use/see this, I'm pulling in my dependencies and dependent files, then after all that 3rd party code loads and before I start running my own code I want to go ahead and turn on uncaught exceptions -- so in this case I would be in control of the underlying code. Are you picturing something different?

**Comment by justingrant** [comment]
The main challenge is that re-building an app to add the code change can be time-consuming or may require you to set up a build environment. If a build takes 5 minutes, then it's much easier to change a debug setting than the run a full re-build to add a function call that's only used for that debug session, and then another 5-minute build to remove it. 

It could in theory be good to make this a best practice that all apps use after their own code starts up, but unfortunately with many frameworks "after their own code starts up" is not always obvious, and will get even less obvious as lazy-loading becomes more popular. 

TL;DR - I still think a time-based debug setting would handle this best. I also like the runtime-code way to trigger this, but I think "both" would be the best solution.

**Comment by JeremyFunk** [comment]
Any updates on this?

**Comment by connor4312** [comment]
I believe [conditional exception breakpoints](https://github.com/microsoft/vscode/issues/104453#issuecomment-754267066) are the better replacement for this feature request.

**Comment by moltenform** [comment]
I was encountering the same issue: wanting to break on exceptions in my code but not on the many expected exceptions occurring during app start. @connor4312's workaround did help me.

For example, for my nodejs project,

I added the line `global.shouldBreakOnExceptions = true` at the start of my tests.

Then in VSCode I enabled "Break on Caught Exceptions" but with the condition `global.shouldBreakOnExceptions`.
